### ISSUE TITLE
AutobahnPython doesn't print critical error messages by default

### ISSUE THREAD

**Issue by jcelliott** [tavendo/AutobahnPython#446]
While debugging several issues with Autobahn (including #445), I realized that critical information that would help track down the problem was being hidden by default. Errors like these should be visible to the user by default:

```
Failing WAMP-over-WebSocket transport: code = 1011, reason = 'WAMP Internal Error (resolve() takes exactly 2 arguments (1 given))'
WAMP-over-WebSocket transport lost: wasClean = False, code = 1006, reason = 'connection was closed uncleanly (I failed the WebSocket connection by dropping the TCP connection)'
```

Instead, I didn't see them until I found the `debug`, `debug_wamp`, and `debug_app` parameters in the API Reference. It's also not clear what the distinction is between them, or why they should all exist. With the simple test script I used in #445 there is no output at all from Autobahn without specifying `debug_wamp=True` and the application simply appears to hang.

**Comment by meejah** [comment]
I think it would be best to eliminate ``debug_app`` entirely -- it's actually not really used any more in Autobahn.

``debug`` and ``debug_wamp`` both provide verbose information about sent and received wamp messages, and call-tracing type information but ``debug_wamp`` is only used in a few spots in autobahn/wamp/websocket.py -- I wonder if we can just collapse these two together into ``debug``. The semantics of this should be to hide messages only intended to be used for debugging Autobahn itself.

@oberstet what do you think?

For your first message quoted above, I think there are several cases of exceptions being hidden behind ``if self.debug_wamp:`` when they shouldn't be. At the very least, I think we should switch around the defaults and provide a ``hide_internal_errors`` or similar...? However, it's tough to imagine a situation where you *wouldn't* want to see the stack-traces from such errors -- my preference would be to simply print/log all such stack-traces for any exception caught and not re-raised. These exceptions represent either bugs in Autobahn itself or in user-overridden code and in either case seeing the exception is usually preferable (and usually expected by Python programmers). I suppose another solution here would be to use ``debug_app`` for these cases, but make it default to ``True``?

As to your *second* message above, it's a bit tough: you can get/log that information by overriding ``onClose`` on your session, but if you *don't* do this it's hard to tell the session has been closed...(however, that wouldn't be an issue here if the exception wasn't hidden in the first place).

**Comment by jcelliott** [comment]
> you can get/log that information by overriding onClose on your session

Did you mean by implementing `onLeave`? `onClose` only gives me a boolean `wasClean` value. `onLeave` gives me the following CloseDetails object:

```
CloseDetails(reason = wamp.close.transport_lost, message = 'WAMP transport was lost without closing the session before'')
```

That was one of the first things I tried while debugging, and it led me the wrong direction because this sounds (to me) like the connection was dropped by the other end (until you see the context of "I failed the WebSocket connection by dropping the TCP connection")

In any case, collapsing the debug* arguments seems like a good idea, as does printing exception stack-traces.

**Comment by meejah** [comment]
Yes you're right: ``onLeave``.

Also I agree that message is a little misleading; perhaps that can be improved also when the client-side is dropping the connection due to (internal) errors.

**Comment by oberstet** [comment]
@jcelliott sorry for the hassles=(

Getting rid of `debug_app`: yes, I think we should do that. Ideally, any exception/issue resulting from user code should not be logged or swallowed, but pop up in user code or be treated there.

Collapsing `debug` and `debug_wamp`: I agree. The historic reason for this is: `debug` turns on incredibly verbose WebSocket level logging (bytes on wire etc), and `debug_wamp` is also very verbose, but only logs the parsed WAMP messages. But both do log a lot (and hence can't be used with significant app messaging).

We should probably go even further: get rid of `debug` flags in all those ctors altogether and follow how CB does logging. @meejah  What do you think?

**Comment by jcelliott** [comment]
FYI, going from 0.10.4 to 0.10.5 it looks like those low level websocket messages aren't being printed anymore even with `debug` on.

**Comment by meejah** [comment]
Yes, I think improving logging in Autobahn similar to the Crossbar improvements will be a good step forward, especially for exceptions etc. I presume you mean putting most (all) of the ``.debug/.debug_wamp`` into a ``trace`` log-level so they're easier to control etc? Also of course changing some of the currently-eaten exceptions to ``.failure`` or ``.critical`` logs (with no ``if``).

I think we should do the logging improvements in a separate branch/PR so that @hawkowl can at least take a look as well as she's a lot more versed in the new-Twisted-logger etc.

**Comment by oberstet** [comment]
@meejah that's a good plan. yeah, the really verbose logging of bytes or messages on the wire could move to `trace` level, while probably other less verbose things could be logged at level `debug`.

**Comment by hawkowl** [comment]
I'll take point on this.

**Comment by singlecheeze** [comment]
I ran into this same issue and spent a few hours deep diving in code before I found the issue of mis placed args.

Needed this for shared registrations:
``` python
self.register(self.parse, 'mono.parse', options=RegisterOptions(invoke='roundrobin'))
```
Instead of this:
``` python
self.register('mono.parse', self.parse, options=RegisterOptions(invoke='roundrobin'))
python
